examples/dup2_to_replace_stdio.rs

```diff
@@ -0,0 +1,40 @@
+//! This is an example of how to use `dup2` to replace the stdin and stdout file
+//! descriptors.
+
+#[cfg(not(windows))]
+fn main() {
+    use io_lifetimes::AsFilelike;
+    use rustix::io::{dup2, pipe};
+    use std::io::{BufRead, BufReader, Write};
+    use std::mem::forget;
+
+    let (reader, writer) = pipe().unwrap();
+    let (stdin, stdout) = unsafe { (rustix::io::take_stdin(), rustix::io::take_stdout()) };
+    dup2(&reader, &stdin).unwrap();
+    dup2(&writer, &stdout).unwrap();
+    forget(stdin);
```

Comment:
Hmm, why does `dup2` take `&OwnedFd` instead of consuming an `OwnedFd`?

Reply:
It's a good question :-). It could take an `OwnedFd` and return an `OwnedFd` with the same value, and it'd be almost exactly the same. But there are a few situations where taking a `&OwnedFd` works where consuming a `OwnedFd` doesn't. For example:

```rust
struct MyType {
    fd: OwnedFd
}
impl MyType {
    fn frob(&mut self) -> io::Result<()> {
        self.fd = dup2(&something, self.fd)
    }
}
```
This doesn't work, because `self` is borrowed, so the type checker won't allow `self.fd` to be consumed. But taking an `&OwnedFd` means you can do `dup2(&something, &self.fd)`, which does work.

Reply:
Couldn't this be fixed by adding something like
`fn dup2_to(src: &Fd, dest: &mut OwnedFd) -> io::Result<()>` ?

Reply:
Do you mean in addition to having a `dup2` that takes an `OwnedFd` by value? We could do that, but it's unclear what advantage that would have over the current `dup2` API.

Reply:
Right, in addition.  Dunno, not a big deal right now.  I know this is just a test case.  But `mem::forget()` to me is usually a big red flag and at a minimum deserves a comment.

Reply:
I've opened https://github.com/bytecodealliance/rustix/pull/207 to add comments.

The `mem::forget` would still be needed, even with a `dup2` function that consumes the `OwnedFd`, because such a function would return a new `OwnedFd`, and then we'd have to `mem::forget` that here.

I think the real tricky part of this example isn't the `dup2` per se, but the `stdin` and `stdout`. We don't want the behavior of tracking their ownership and closing them when they're dropped. We just want to install them as new ambient file descriptors for `println` et al to use.